storage/src/journal/contiguous/variable.rs

```diff
@@ -0,0 +1,1409 @@
+            // 1. We completely pruned the data journal but crashed before pruning
+            //    the offsets journal.
+            // 2. The data journal was never opened.
+            if let Some(oldest) = offsets.oldest_retained_pos().await? {
+                if oldest < size {
+                    // Offsets has unpruned entries but data is gone - repair by pruning
+                    info!("crash repair: pruning offsets to {size} (prune-all crash)");
+                    offsets.prune(size).await?;
+                    offsets.sync().await?;
+                }
+            }
+
+            return Ok((size, size));
+        }
+
+        // === Handle non-empty data journal case ===
+        let (data_oldest_pos, data_size) = {
+            // Data exists -- count items
+            let first_section = *data.blobs.first_key_value().unwrap().0;
+            let last_section = *data.blobs.last_key_value().unwrap().0;
+
+            let oldest_pos = first_section * items_per_section;
+
+            // Invariant 1 on `Variable` guarantees that all sections except possibly the last
+            // are full. Therefore, the size of the journal is the number of items in the last
+            // section plus the number of items in the other sections.
+            let size = (last_section * items_per_section) + items_in_last_section;
+            (oldest_pos, size)
+        };
+        assert_ne!(
+            data_oldest_pos, data_size,
+            "data journal expected to be non-empty"
+        );
+
+        // Align pruning state. We always prune the data journal before the offsets journal,
+        // so we validate that invariant and repair crash faults or detect corruption.
+        match offsets.oldest_retained_pos().await? {
+            Some(oldest_retained_pos) if oldest_retained_pos < data_oldest_pos => {
+                // Offsets behind on pruning -- prune to catch up
+                info!("crash repair: pruning offsets journal to {data_oldest_pos}");
+                offsets.prune(data_oldest_pos).await?;
+            }
+            Some(oldest_retained_pos) if oldest_retained_pos > data_oldest_pos => {
+                return Err(Error::Corruption(format!(
+                    "offsets oldest pos ({oldest_retained_pos}) > data oldest pos ({data_oldest_pos})"
+                )));
+            }
+            Some(_) => {
+                // Both journals are pruned to the same position.
+            }
+            None if data_oldest_pos > 0 => {
+                // Offsets journal is empty (size == oldest_retained_pos).
+                // This can happen if we pruned all data, then appended new data, synced the
+                // data journal, but crashed before syncing the offsets journal.
+                // We can recover if offsets.size() matches data_oldest_pos (proper pruning).
+                let offsets_size = offsets.size().await?;
+                if offsets_size != data_oldest_pos {
```

Comment:
idk why the intuition around this logic escaped me a few days ago but I now "get it" (we only allow the pruning to get "one sync ahead" or else consider it corrupt)